hazelcast/src/test/java/com/hazelcast/test/annotation/QuickTest.java

```diff
@@ -16,10 +16,26 @@
 
 package com.hazelcast.test.annotation;
 
+import org.junit.runners.model.FrameworkMethod;
+
+import java.text.MessageFormat;
+import java.time.Duration;
+
 /**
- * Annotates quick tests which are fast enough for the PR builder.
+ * Annotates quick tests which are fast enough (i.e. execution sub-{@link #EXPECTED_RUNTIME_THRESHOLD} per test) for the PR builder.
  * <p>
  * Will be executed in the PR builder and for code coverage measurements.
+ * <p>
+ * The opposite of a {@link SlowTest}
  */
 public final class QuickTest {
+    public static final Duration EXPECTED_RUNTIME_THRESHOLD = Duration.ofMinutes(1);
+
+    public static void logMessageIfTestOverran(FrameworkMethod method, float tookSeconds) {
+        if (tookSeconds > QuickTest.EXPECTED_RUNTIME_THRESHOLD.getSeconds()) {
+            System.err.println(MessageFormat.format(
+                    "{0} is annotated as a {1}, expected to complete within {2} seconds - but took {3} seconds",
+                    method.getName(), QuickTest.class.getSimpleName(), EXPECTED_RUNTIME_THRESHOLD.getSeconds(), tookSeconds));
+        }
+    }
```

Comment:
We have valid cases where we want to run a test taking more than 1 minute in PR builder. This will cause noise for such tests.

First I thought we can add an annotation parameter to `QuickTest` to mark such tests, but it is a class, not an annotation. So maybe introduce a new category or completely separate annotation which we can be used to mark long running tests which should run in PR builder and it would be checked here and we wouldn't print the log for such cases.

Wdyt? @vbekiaris @k-jamroz  your thoughts?

Reply:
maybe not report warning for `@Category({QuickTest.class, SlowTest.class})` (abusing `SlowTest` meaning)? 
BTW, didn't we consider removing `QuickTest.class` entirely, as it is (was before this PR) a de-facto no-op.

Reply:
IMO it would be better to have an easy way to distinguish and find exceptional tests, so I'd prefer dedicated category or annotation

Reply:
> We have valid cases where we want to run a test taking more than 1 minute in PR builder. 

Did you have a specific test in mind? Just to check I haven't disabled it.

> This will cause noise for such tests

> a warning that nobody (?) will read

The primary case I was thinking of was where, while a feature is developed, the test starts as a `QuickTest` - but as more cases are identified, it gets slower and slower but no-one considers it's crossed the threshold. 

> First I thought we can add an annotation parameter to `QuickTest` to mark such tests, but it is a class, not an annotation. So maybe introduce a new category or completely separate annotation which we can be used to mark long running tests which should run in PR builder and it would be checked here and we wouldn't print the log for such cases.

That's a nice solution but I worry it's an extra layer of complexity - were probably better off not warning at all in that case.

> but if we really want to enforce the rule, we should fail the tests

The aim of this change is to make the PR builder nicer... I don't think adding more failures would make me very popular.


It doesn't sound like this check is well-received - I'm leaning towards removing it.

Reply:
>> a warning that nobody (?) will read
> The primary case I was thinking of was where, while a feature is developed, the test starts as a QuickTest - but as more cases are identified, it gets slower and slower but no-one considers it's crossed the threshold.

Yes, I also see this phenomenon. During intensive development of some feature it is beneficial to have its tests running in the PR builder, but when the feature is mature and the development stops, and it is not a core feature that everything else depends on, the benefit of running all those tests in PR builder is much smaller, even if they take <1 minute each.

> > but if we really want to enforce the rule, we should fail the tests
> 
> The aim of this change is to make the PR builder nicer... I don't think adding more failures would make me very popular.

Yes, I agree that in current situation this would cause more harm than good.
 
> It doesn't sound like this check is well-received - I'm leaning towards removing it.

I am very in favour of making PR builder faster, and anything that leads to that is good. Of course I'd like to run all SQL and Jet tests in PR builds but this is infeasible.
BTW, I just had another idea: maybe each module should get a budget of time in PR builder (eg. 5min for SQL, 5 min for Jet, 10min for Core, etc) and each team could allocate it as they see fit: be it one big test or many small tests.

There is balance to find with moving tests from PR builds to nightly. We have currently imperfect process of handling failures in nightly test failures, so there is a compromise between speed of PR builds and speed and cost of fixing regressions. Another consideration is if we should run more tests for community contributions before merging, or which test might be important (mandatory?) for such PRs. This is much more complicated that just test runtime, but it is a first step.

Reply:
> Did you have a specific test in mind? Just to check I haven't disabled it.

E.g. Jet extensions tests (StreamKafkaPTest in your list) - each extension should be tested end to end for at least some happy path and common failure scenarios.

We should focus on tests in `hazelcast` module, which do not have `ParallelJVMTest` annotation - they take up their full time.

Other tests run in parallel in IIRC 8 threads so if you moving 80s test will give you only 10s on whole PR builder time.

The extension modules are also built in parallel, but there is a critical path, which used to be:

`hazelcast` <- `hazelcast-jet-kafka` <- `hazelcast-sql`

everything else outside of these modules doesn't matter, but maybe this changed - there is `hazelcast-jet-mongodb`, not sure if this one takes longer than kafka.

Reply:
> BTW, didn't we consider removing `QuickTest.class` entirely, as it is (was before this PR) a de-facto no-op.

I don't think it's a good idea. When you see a test category explicitly, at least you may think to re-classify the test if it consumes too much time. With implicit default, we will have more heavyweight tests added to the PR builder by mistake, due to the implicitness of test category.

Reply:
> Actually, failing your test if it doesn't have a defined category is a good idea. WDYT @JackPGreen?

I think it's good to have consistency - i.e. everywhere uses `QuickTest` or nowhere - but I think there's going to be a lot of consistency to iron out in the process.